src/server/middleware/csrf/csrf.go

```diff
@@ -0,0 +1,90 @@
+package csrf
+
+import (
+	"fmt"
+	"github.com/goharbor/harbor/src/common/utils"
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/core/config"
+	"github.com/goharbor/harbor/src/server/middleware"
+	"github.com/gorilla/csrf"
+	"net/http"
+	"os"
+	"strings"
+	"sync"
+)
+
+const (
+	csrfKeyEnv  = "CSRF_KEY"
+	tokenHeader = "X-Harbor-CSRF-Token"
+	tokenCookie = "__csrf"
+)
+
+var (
+	once    sync.Once
+	protect func(handler http.Handler) http.Handler
+)
+
+// attachToken makes sure if csrf generate a new token it will be included in the response header
+func attachToken(w http.ResponseWriter, r *http.Request) {
+	if t := csrf.Token(r); len(t) > 0 {
+		http.SetCookie(w, &http.Cookie{
+			Name:     tokenCookie,
+			Secure:   true,
+			Value:    t,
+			Path:     "/",
+			SameSite: http.SameSiteStrictMode,
+		})
+	} else {
+		log.Warningf("token not found in context, skip attaching")
+	}
+}
+
+func handleError(w http.ResponseWriter, r *http.Request) {
+	attachToken(w, r)
+	http.Error(w, fmt.Sprintf("%s - %s",
```

Comment:
Use https://github.com/goharbor/harbor/blob/master/src/server/error/error.go#L49 instead, so that we can control the error format in one place

Reply:
Yes, definitely!